File: days/day_25/main.py
# -*- coding: utf-8 -*-
import time
import os

def transform(subject, loop):
	value = 1
	modulus = 20201227
	while loop >= 1:
		if loop % 2 == 1:	
			value *= subject
			value %= modulus
		
		subject *= subject
		subject %= modulus

		loop//=2
	return value


def find_loop(key, subject):
	loop = 1
	while transform(subject, loop) != key:
		loop += 1
	return loop


def part_one(card_key, door_key, subject):
	"""
	Using Sympy's discrete log:

	from sympy.ntheory.residue_ntheory import discrete_log

	card_loop = discrete_log(20201227, card_key, subject)
	door_loop = discrete_log(20201227, door_key, subject)
	"""

	card_loop = find_loop(card_key, subject)
	door_loop = find_loop(door_key, subject)

	return transform(transform(subject, door_loop), card_loop)


def main():
	start_time = time.time()

	with open(os.path.dirname(__file__) + '/input.txt', 'r') as data:
		card_key = int(data.readline())
		door_key = int(data.readline())

		part_one_ans = part_one(card_key, door_key, 7)
		part_two_ans = '⭐️'

		print('Day 25  ({:,.3f}s)'.format(time.time()-start_time))
		print('  Part 1: {}'.format(part_one_ans))
		print('  Part 2: {}'.format(part_two_ans))


if __name__ == "__main__":
		main()

File: days/day_25/main.ml
open Core

let file = "input.txt"

let transform subject loop =
	let modulus = 20201227 in
	let rec transform subject loop value =
		if loop = 0 
		then value
		else if loop mod 2 = 1 
		then transform ((subject * subject) mod modulus)
									 (loop / 2)
									 ((value * subject) mod modulus)
		else transform ((subject * subject) mod modulus)
									 (loop / 2)
									 value
	in
	transform subject loop 1
;;

let find_loop key subject =
	let rec check loop =
		if transform subject loop = key then loop
		else check (loop+1)
	in
	check 1
;;

let part_one card_key door_key subject =
	let card_loop = find_loop card_key subject in
	let door_loop = find_loop door_key subject in
	transform (transform subject door_loop) card_loop
;;

let () = 
	let start_time = Unix.gettimeofday () in
	let input = In_channel.create file in
	let card_key = int_of_string (Option.value_exn (In_channel.input_line input)) in
	let door_key = int_of_string (Option.value_exn (In_channel.input_line input)) in
	let part_one_ans = part_one card_key door_key 7 in
	let part_two_ans = "⭐️" in
	printf "Day 25 (%.3fs)\n" ((Unix.gettimeofday ()) -. start_time);
	printf "  Part 1: %d\n" part_one_ans;
	printf "  Part 2: %s\n" part_two_ans
;;
